go/simple/v2/ampli/ampli.go

```diff
@@ -0,0 +1,388 @@
+	}
+}
+
+// EventWithAllProperties
+//
+// [View in Tracking Plan]: https://data.amplitude.com/test-codegen/Test%20Codegen/events/main/latest/Event%20With%20Array%20Types
+//
+// Event with all properties
+//
+// Owner: Test codegen
+//
+// Params
+//	- required_array: Event 2 Property - Array
+//	- required_boolean: Event 2 Property - Boolean
+//	- required_enum: Event 2 Property - Enum
+//	- required_integer: Event 2 Property - Integer
+//	- required_number: Event 2 Property - Number
+//	- required_string: Event 2 Property - String
+//	- optional_string: Event 2 Property - Optional String
+type EventWithAllProperties struct {
+	RequiredArray   []string
+	RequiredBool    bool
+	RequiredEnum    RequiredEnum
+	RequiredInteger int
+	RequiredNumber  float64
+	RequiredString  string
+	OptionalString  string
+	OptionalBool    bool
+	optionals       map[string]bool
+}
+
+type RequiredEnum string
+
+const (
+	RequiredEnumEnum1 RequiredEnum = "Enum1"
+	RequiredEnumEnum2 RequiredEnum = "Enum2"
+)
+
+func NewEventWithAllProperties(requiredArray []string, requiredBool bool, requiredEnum RequiredEnum, requiredInteger int, requiredNumber float64, requiredString string) *EventWithAllProperties {
+	return &EventWithAllProperties{
+		RequiredArray:   requiredArray,
+		RequiredBool:    requiredBool,
+		RequiredEnum:    requiredEnum,
+		RequiredInteger: requiredInteger,
+		RequiredNumber:  requiredNumber,
+		RequiredString:  requiredString,
+		optionals:       map[string]bool{},
+	}
+}
+
+func (stronglyTypedEvent *EventWithAllProperties) SetOptionalString(optionalString string) *EventWithAllProperties {
+	stronglyTypedEvent.OptionalString = optionalString
+	stronglyTypedEvent.optionals["OptionalString"] = true
+
+	return stronglyTypedEvent
+}
+
+func (stronglyTypedEvent *EventWithAllProperties) SetOptionalBool(optionalBool bool) *EventWithAllProperties {
+	stronglyTypedEvent.OptionalBool = optionalBool
+	stronglyTypedEvent.optionals["OptionalBool"] = true
+
+	return stronglyTypedEvent
+}
+
+func (stronglyTypedEvent EventWithAllProperties) ToEvent() amplitude.Event {
+	event := amplitude.Event{
+		EventType: "Event With All Properties",
+		EventProperties: map[string]interface{}{
+			"requiredArray":   stronglyTypedEvent.RequiredArray,
+			"requiredBool":    stronglyTypedEvent.RequiredBool,
+			"requiredEnum":    stronglyTypedEvent.RequiredEnum,
+			"requiredInteger": stronglyTypedEvent.RequiredInteger,
+			"requiredNumber":  stronglyTypedEvent.RequiredNumber,
+			"requiredString":  stronglyTypedEvent.RequiredString,
+		},
+	}
+
+	if _, present := stronglyTypedEvent.optionals["OptionalString"]; present {
+		event.EventProperties["optionalString"] = stronglyTypedEvent.OptionalString
+	}
+
+	if _, present := stronglyTypedEvent.optionals["OptionalBool"]; present {
+		event.EventProperties["optionalBool"] = stronglyTypedEvent.OptionalBool
+	}
+
+	return event
+}
+
+type Ampli struct {
+	Disabled bool
+	Client   amplitude.Client
+	mutex    sync.Mutex
```

Comment:
`sync.RWMutex`?

Reply:
It's better to use `RWMutex`: 
- use `RLock()` and `RUnlock()` when reading, `InitializedAndEnabled()`
- keep `Lock()`and `Unlock() when writing, `ampli.Load()` and `ampli.Shutdown()`